### PR TITLE
Final HLint pass

### PR DESCRIPTION
**Pull request by lwm**
~~I wasn't sure about the duplications (which I suppressed - will revisit).~~ Ignored.

~~Just dropping this one here for now.~~

~~Needs a rebase over changes in #2688 to get a clean Travis run.~~

Next is #2591!

### CODE REVIEW

File: HLint.hs [hunk 1]
```diff
@@ -1,7 +1,33 @@
 -- HLint configuration file
 
+module HLint.HLint where
+
+-- mimic default HLint.hs
+-- https://github.com/ndmitchell/hlint/blob/master/data/HLint.hs
+import "hint" HLint.Default
```
Comment (mgsloan) [inline HLint.hs]: While it's in the default HLint.hs, why are these package qualified imports that look like they come from https://hackage.haskell.org/package/hint ?  Seems strange!

File: src/Stack/Build/Execute.hs [hunk 2]
```diff
@@ -586,9 +584,7 @@ executePlan' installedMap0 targets plan ee@ExecuteEnv {..} = do
             -- TODO it probably makes more sense to use a lock for test suites
             -- and just have the execution blocked. Turning off all concurrency
             -- on finals based on the --test option doesn't fit in well.
-            if boptsTests eeBuildOpts
-                then concurrentTests
-                else True
+            not (boptsTests eeBuildOpts) || concurrentTests
```
Comment (mgsloan) [inline src/Stack/Build/Execute.hs]: I like the old version better, clearer to me.

File: src/Stack/Build/Execute.hs [hunk 3]
```diff
@@ -647,7 +643,7 @@ toActions installedMap runInBase ee (mbuild, mfinal) =
                 [ Action
                     { actionId = ActionId taskProvides ATBuild
                     , actionDeps =
-                        Set.map (\ident -> ActionId ident ATBuild) (tcoMissing taskConfigOpts)
+                        Set.map (`ActionId` ATBuild) (tcoMissing taskConfigOpts)
```
Comment (mgsloan) [inline src/Stack/Build/Execute.hs]: I also find the old version more readable here.

File: src/Stack/Build/Execute.hs [hunk 4]
```diff
@@ -659,7 +655,7 @@ toActions installedMap runInBase ee (mbuild, mfinal) =
                     [Action
                         { actionId = ActionId taskProvides ATBuildFinal
                         , actionDeps = addBuild ATBuild
-                            (Set.map (\ident -> ActionId ident ATBuild) (tcoMissing taskConfigOpts))
+                            (Set.map (`ActionId` ATBuild) (tcoMissing taskConfigOpts))
```
Comment (mgsloan) [inline src/Stack/Build/Execute.hs]: I also find the old version more readable here.

File: src/Stack/Setup.hs [hunk 5]
```diff
@@ -555,9 +555,7 @@ getGhcBuild menv = do
                             | hasgmp4 && hastinfo5 -> ["gmp4"]
                             | otherwise -> []
                     pieComponents =
-                        if noPie
-                            then ["nopie"]
-                            else []
+                        ["nopie" | noPie]
```
Comment (mgsloan) [inline src/Stack/Setup.hs]: I also find the old version more readable here.

File: src/Stack/Types/FlagName.hs [hunk 6]
```diff
@@ -54,7 +54,7 @@ newtype FlagName =
   FlagName Text
   deriving (Typeable,Data,Generic,Hashable,Store,NFData)
 instance Eq FlagName where
-    x == y = compare x y == EQ
+    x == y = x == y
```
Comment (mgsloan) [inline src/Stack/Types/FlagName.hs]: This will infinite loop.  You can't trust hlint to check semantics!

File: src/test/Stack/StoreSpec.hs [hunk 7]
```diff
@@ -48,7 +48,7 @@ instance (Monad m, Serial m a, Ord a) => Serial m (Set a) where
 
 addMinAndMaxBounds :: forall a. (Bounded a, Eq a) => [a] -> [a]
 addMinAndMaxBounds xs =
-    (if (minBound :: a) `notElem` xs then [minBound] else []) ++
+    [minBound | (minBound :: a) `notElem` xs] ++
```
Comment (mgsloan) [inline src/test/Stack/StoreSpec.hs]: I also find the old version more readable here.

File: src/Stack/Nix.hs [hunk 8]
```diff
@@ -115,7 +115,7 @@ runShellAndExit mprojectRoot getCompilerVersion getCmdArgs = do
                               ,"STACK_IN_NIX_EXTRA_ARGS = stackExtraArgs; "
                               ,"} \"\""]]
                     -- glibcLocales is necessary on Linux to avoid warnings about GHC being incapable to set the locale.
-         fullArgs = concat [if pureShell then ["--pure"] else []
+         fullArgs = concat [["--pure" | pureShell]
```
Comment (mgsloan) [inline src/Stack/Nix.hs]: I also find the old version more readable here.

File: src/Stack/Package.hs [hunk 9]
```diff
@@ -1080,7 +1080,7 @@ findCandidate dirs exts name = do
         -- The standard canonicalizePath does not work for this case
         p <- parseCollapsedAbsFile (toFilePath x FilePath.</> y)
         exists <- doesFileExist p
-        return $ if exists then [p] else []
+        return [p | exists]
```
Comment (mgsloan) [inline src/Stack/Package.hs]: I also find the old version more readable here.

**Comment by mgsloan** [comment]
Thanks for working on this!  Glad this is the final HLint pass, I think our attention is better spent elsewhere.  Changing code can always introduce bugs (like the `==` thing), I would rather make changes to the code that are feature additions and enhancements.  Cleanups are good too, but a fair number of these changes are minor downgrades in clarity.

We are optimizing for code clarity, not code concision or what hlint thinks.

File: src/Stack/Types/FlagName.hs (hunk 6 above)
Comment (lwm) [inline src/Stack/Types/FlagName.hs]: Oof, I thought I ignore those ones. Will do.

File: HLint.hs (hunk 1 above)
Comment (lwm) [inline HLint.hs]: It looks strange to me as well but this is how HLint rolls.

**Comment by lwm** [comment]
Thanks for the review @mgsloan. I've addressed all above comments by reverting the changes and adding those type of hints to the ignore file.

**Comment by mgsloan** [comment]
Cool, thanks @lwm!

**Pull request merged**